### DISCUSSION TITLE
SmartMeter Interface: txPeriod=100ms and Modbus

### DISCUSSION THREAD

**Discussion by shikousa** [arendst/Tasmota#20032] [Q&A]
Hello,

I have seen that according to the templates a lot of Modbus meters (e.g. SDM530, SDM72D, EM340 ...) are using a txPeriod of 100ms or 200ms. Does this mean that Tasmota is requesting every 100ms/200ms new values from the energy meter? Isn't this a little bit too fast.

The fastest MQTT period you can set is 10s and even if you would send on request I think it makes no sense to send faster than 3s.
I think a txPeriod of 3000ms should therefore be more than enough? What is your opinion?

@gemu2015 What do you think? If I understand the code correct you send the <cmdTelegram> even if it contains several Modbus slave device IDs every <txPerdiod>? E.g. if you have 3 slaves does it then makes sense to define 1 or 3 energy definitions in Tasmota?

**Comment by gemu2015** [comment]
that depends on how many values you fetch and if you fetch more than one register at once.
most descriptor examples fetch single registers and therefore the whole set takes n * 200 ms
you are free to increase the interval as much as you like.
according the lower limit:
it seems that most meters can respond in 100 ms but some need 200, so 200 is a good value to be sure. 

some users need a very fast response because they use control loops for solar inverters 
in this case MQTT isn't fast enough either and the loop is implemented as a script


a meter definition in this driver is assigned to one bus each, so multiple meters on the same bus must be defined in one line

here a 2 examples
 
```
; single float register request, takes 9*500 ms 
>M 1  
+1,[192.168.188.106],m,0,502,SDM630,0,5,01040000,01040002,01040004,01040006,01040008,0104000a,0104000c,0104000e,01040010  
1,010404ffffffff@i0:1,Voltage P1,V,Voltage_P1,2  
1,010404ffffffff@i1:1,Voltage P2,V,Voltage_P2,2  
1,010404ffffffff@i2:1,Voltage P3,V,Voltage_P3,2  
1,010404ffffffff@i3:1,Current P1,A,Current_P1,2  
1,010404ffffffff@i4:1,Current P2,A,Current_P2,2  
1,010404ffffffff@i5:1,Current P3,A,Current_P3,2  
1,010404ffffffff@i6:1,Active Power P1,W,Power_P1,2  
1,010404ffffffff@i7:1,Active Power P2,W,Power_P2,2  
1,010404ffffffff@i8:1,Active Power P3,W,Power_P3,2  
#  
```

```
; multiple float register request, 9 at once, takes 500 ms
>M 1  
+1,[192.168.188.106],m,0,502,SDM630,0,5,r010400000012
1,010424ffffffff@i0:1,Voltage P1,V,Voltage_P1,2  
1,010424x4ffffffff@i0:1,Voltage P2,V,Voltage_P2,2  
1,010424x8ffffffff@i0:1,Voltage P3,V,Voltage_P3,2  
1,010424x12ffffffff@i0:1,Current P1,A,Current_P1,2  
1,010424x16ffffffff@i0:1,Current P2,A,Current_P2,2  
1,010424x20ffffffff@i0:1,Current P3,A,Current_P3,2  
1,010424x24ffffffff@i0:1,Active Power P1,W,Power_P1,2  
1,010424x28ffffffff@i0:1,Active Power P2,W,Power_P2,2  
1,010424x32ffffffff@i0:1,Active Power P3,W,Power_P3,2  
#
```

**Comment by shikousa** [comment]
Thanks for the info. Ok, that means every txPeriod you send a new command, but not all commands at once. If you have three energy meters then the definition would look like this. In this example every energy meter is then called every 1500ms? If one energy meter goes offline, will it then be ignored, is there something like a repeat or timeout?

```
; 3 meters, every energy meter is polled after 3x500ms = 1500ms; Slave ID = 1,2,3
>M 1  
+1,[192.168.188.106],m,0,502,SDM630,0,5,r010400000012,r020400000012,r030400000012
```

**Reply by gemu2015** [reply]
yes the above meter definition is correct for scanning 3 meters
an offline meter simply will not respond.
there is no repeat for a single request since the meter is frequently requested at every 1500 ms in your example.
a modbus meter going offline is however an extremely rare case at least with serial modbus
however you may detect an offline meter with the smlv[x] (sml valid) flag in the script.

**Comment by shikousa** [comment]
Is it also possible to define something like an update interval of e.g. 60s for Modbus? That means every 60s I want to send all commands. At the moment it seems that you have a continous sending of the commands every txPeriod with no wait time after the last command. That produces a lot of load on the bus and also a lot of data which is not required.

**Comment by shikousa** [comment]
@gemu2015 Can you please give me a short feedback if I can add something like a delay time for Modbus?

**Reply by gemu2015** [reply]
no delay implemented
